README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# mrIML: Multivariate (multi-response) interpretable machine learning <a href='https://nfj1380.github.io/mrIML/index.html'><img src="man/figures/logo.png" align="right" height="120"/></a>

<!-- badges: start -->

![GitHub R package
version](https://img.shields.io/github/r-package/v/nfj1380/mrIML?logo=github&logoColor=%2300ff37&style=flat-square)
![GitHub
contributors](https://img.shields.io/github/contributors/nfj1380/mrIML?style=flat-square)
![GitHub all
releases](https://img.shields.io/github/downloads/nfj1380/mriml/total?style=flat-square)
![GitHub last
commit](https://img.shields.io/github/last-commit/nfj1380/mrIML?style=flat-square)
[![R-CMD-check](https://github.com/nfj1380/mrIML/workflows/R-CMD-check/badge.svg)](https://github.com/nfj1380/mrIML/actions)
<!-- badges: end -->

This package aims to enable users to build and interpret multivariate
machine learning models harnessing the tidyverse (tidy model syntax in
particular). This package builds off ideas from Gradient Forests ([Ellis
et al
2012](https://esajournals-onlinelibrary-wiley-com.prox.lib.ncsu.edu/doi/full/10.1890/11-0252.1)),
and ecological genomics [Fitzpatrick and Keller,
2014](https://onlinelibrary.wiley.com/doi/abs/10.1111/ele.12376).

This package can be of use for any multi-response machine learning
problem, but was designed to handle data common to community ecology
(site by species data) and ecological genomics (individual or population
by SNP loci).

## Recent mrIML publications

1.  Fountain-Jones, N. M., Kozakiewicz, C. P., Forester, B. R.,
    Landguth, E. L., Carver, S., Charleston, M., Gagne, R. B.,
    Greenwell, B., Kraberger, S., Trumbo, D. R., Mayer, M., Clark, N.
    J., & Machado, G. (2021). MrIML: Multi-response interpretable
    machine learning to model genomic landscapes. Molecular Ecology
    Resources, 21, 2766– 2781. <https://doi.org/10.1111/1755-0998.13495>

2.  Sykes, A. L., Silva, G. S., Holtkamp, D. J., Mauch, B. W., Osemeke,
    O., Linhares, D. C.L., & Machado, G. (2021). Interpretable machine
    learning applied to on-farm biosecurity and porcine reproductive and
    respiratory syndrome virus. Transboundary and Emerging Diseases, 00,
    1– 15. <https://doi.org/10.1111/tbed.14369>

## Installation

Install the stable version of the package:

``` r
#install.packages("devtools")
#devtools::install_github('nfj1380/mrIML')
library(mrIML)
```

## Quick start

**mrIML** is designed to be used with a single function call or to be
used in an ad-hoc fashion via individual function calls. In the
following section we give an overview of the simple use case. For more
on using each function see the [function
documentation](https://nfj1380.github.io/mrIML/reference/index.html).
The core functions for both regression and classification are:
[`mrIMLpredicts`](https://nfj1380.github.io/mrIML/reference/mrIMLpredicts.html),
[`mrIMLperformance`](https://nfj1380.github.io/mrIML/reference/mrIMLperformance.html),
and
[`mrInteractions`](https://nfj1380.github.io/mrIML/reference/mrInteractions.html),for
plotting and visualization
[`mrVip`](https://nfj1380.github.io/mrIML/reference/mrVip.html),
[`mrFlashlight`](https://nfj1380.github.io/mrIML/reference/mrFlashlight.html),
and[`plot_vi`](https://nfj1380.github.io/mrIML/reference/plot_vi.html).
Estimating the interactions alone can be substantially computationally
demanding depending on the number of outcomes you want to test. The
first step to using the package is to load it as follows.

## Model component

Now all the data is loaded and ready to go we can formulate the model
using tidymodel syntax. In this case we have binary data (SNP
presence/absence at each loci) but the data could also be counts or
continuous (the set\_model argument would be “regression” instead of
“classification”). The user can specify any model from the ‘tidymodel’
universe as ‘model 1’ (see <https://www.tidymodels.org/find/> for
details). However, we have done most of our testing on random forests
(rf) and glms (generalized linear models). Here we will specify a random
forest classification model as the model applied to each response.

``` r
model1 <- 
  rand_forest(trees = 10, mode = "classification", 
              mtry = tune(), 
              min_n = tune()) %>% #100 trees are set for brevity
              set_engine("randomForest")
```

### [`mrIMLpredicts`](https://nfj1380.github.io/mrIML/reference/mrIMLpredicts.html)

This function represents the core functionality of the package and
includes results reporting, plotting and optional saving. It requires a
data frame of X t( the snp data for example) and Y represented by the
covariates or features.

Load example data (cite) data from `{mrIML}`.

``` r
fData <- filterRareCommon (Responsedata,
                           lower=0.4,
                           higher=0.7) 
data <- fData[1:20]
```

## Parallel processing

MrIML provides uses the flexible future apply functionality to set up
multi-core processing. In the example below, we set up a cluster using 4
cores. If you don’t set up a cluster, the default settings will be used
and the analysis will run sequentially.

``` r
# detectCores() #check how many cores you have available. We suggest keeping one core free for internet browsing etc.

cl <- parallel::makeCluster(4)

plan(cluster,
     workers=cl)
```

``` r
#Define set of features
fData <- filterRareCommon (Responsedata,
                           lower=0.4,
                           higher=0.7) 
Y <- fData #For simplicity when comparing
#Define set the outcomes of interest
str(Features) 
```

    ## 'data.frame':    20 obs. of  19 variables:
    ##  $ Grassland       : num  0.07 0.0677 0.1845 0.0981 0.1578 ...
    ##  $ Shrub.Scrub     : num  0.557 0.767 0.524 0.786 0.842 ...
    ##  $ Forest          : num  0.01072 0.030588 0.008615 0.000662 0.000616 ...
    ##  $ HighlyDev       : num  0 0 0.00225 0 0 ...
    ##  $ Urban           : num  0 0 0.00159 0 0 ...
    ##  $ Suburban        : num  0.00357 0.13268 0.01325 0.00119 0 ...
    ##  $ Exurban         : num  0.00622 0.03019 0 0.01906 0 ...
    ##  $ Altered         : num  0.441 0.182 0.114 0.12 0 ...
    ##  $ Distance        : num  1.321 0.492 3.231 5.629 4.739 ...
    ##  $ Latitude        : num  33.8 33.8 33.8 33.8 33.8 ...
    ##  $ Longitude       : num  -118 -118 -118 -118 -118 ...
    ##  $ Age             : int  3 0 3 2 3 3 2 3 3 3 ...
    ##  $ Sex             : int  1 1 1 1 0 0 0 1 1 1 ...
    ##  $ Relatedness.PCO1: num  -0.1194 -0.0389 -0.1618 -0.1811 -0.1564 ...
    ##  $ Relatedness.PCO2: num  -0.1947 -0.0525 -0.321 -0.0827 0.1 ...
    ##  $ Relatedness.PCO3: num  -0.191 -0.0874 0.0541 -0.0627 -0.0111 ...
    ##  $ Relatedness.PCO4: num  0.1117 0.2422 0.0974 0.2129 0.2259 ...
    ##  $ Relatedness.PCO5: num  0.06405 0.0706 0.03514 -0.00084 0.0894 ...
    ##  $ Relatedness.PCO6: num  -0.0432 0.0683 -0.0805 0.2247 -0.055 ...

``` r
#Remove NAs from the feature/predictor data.
FeaturesnoNA<-Features[complete.cases(Features), ]
X <- FeaturesnoNA #For simplicity
#For more efficient testing for interactions (more variables more interacting pairs)
X <- FeaturesnoNA[c(1:3)] #Three features only


yhats <- mrIMLpredicts(X=X, #Features/predictors 
                       Y=Y, #Response data
                       Model=model1, #Specify your model
                       balance_data='no', #Chose how to balance your data 
                       mode='classification', #Chose your mode (classification versus regression)
                       seed = 120) #Set seed

ModelPerf <- mrIMLperformance(yhats=yhats,
                              Model=model1,
                              Y=Y, mode='classification')
ModelPerf[[1]] #Predictive performance for individual responses 
```

    ##    response  model_name           roc_AUC                mcc       sensitivity
    ## 1   env_131 rand_forest                 1  0.666666666666667 0.666666666666667
    ## 2   env_163 rand_forest              0.75               <NA>                 0
    ## 3   env_164 rand_forest                 1                  1                 1
    ## 4   env_167 rand_forest 0.583333333333333 -0.408248290463863 0.666666666666667
    ## 5   env_169 rand_forest                 1  0.666666666666667 0.666666666666667
    ## 6   env_212 rand_forest                 1                  1                 1
    ## 7    env_23 rand_forest             0.875  0.408248290463863                 1
    ## 8    env_24 rand_forest              0.25 -0.166666666666667               0.5
    ## 9    env_41 rand_forest 0.583333333333333  0.166666666666667 0.666666666666667
    ## 10   env_47 rand_forest                 1  0.666666666666667                 1
    ## 11   env_59 rand_forest                 1  0.666666666666667 0.666666666666667
    ## 12    env_8 rand_forest 0.666666666666667               <NA>                 1
    ## 13   env_84 rand_forest                 1  0.666666666666667 0.666666666666667
    ## 14   env_85 rand_forest                 1  0.666666666666667 0.666666666666667
    ## 15   env_86 rand_forest               0.5               <NA>                 0
    ## 16  pol_105 rand_forest             0.875  0.612372435695795              0.75
    ## 17  pol_108 rand_forest              0.75  0.408248290463863               0.5
    ## 18  pol_111 rand_forest                 1                  1                 1
    ## 19  pol_117 rand_forest 0.583333333333333  0.166666666666667 0.666666666666667
    ## 20  pol_132 rand_forest                 1  0.666666666666667                 1
    ## 21  pol_159 rand_forest              0.25 -0.166666666666667               0.5
    ## 22  pol_258 rand_forest              0.25 -0.166666666666667               0.5
    ## 23   pol_30 rand_forest                 1  0.666666666666667                 1
    ## 24  pol_340 rand_forest 0.416666666666667 -0.166666666666667 0.333333333333333
    ## 25  pol_353 rand_forest               0.5               <NA>                 0
    ## 26  pol_366 rand_forest                 1  0.666666666666667 0.666666666666667
    ## 27   pol_87 rand_forest                 1  0.666666666666667                 1
    ## 28   pol_88 rand_forest                 1  0.666666666666667                 1
    ## 29   pol_89 rand_forest                 1  0.666666666666667                 1
    ##          specificity        prevalence
    ## 1                  1 0.421052631578947
    ## 2                  1 0.631578947368421
    ## 3                  1 0.421052631578947
    ## 4                  0 0.421052631578947
    ## 5                  1 0.421052631578947
    ## 6                  1 0.684210526315789
    ## 7                0.5 0.631578947368421
    ## 8  0.333333333333333 0.421052631578947
    ## 9                0.5 0.473684210526316
    ## 10 0.666666666666667 0.473684210526316
    ## 11                 1 0.421052631578947
    ## 12                 0 0.473684210526316
    ## 13                 1 0.421052631578947
    ## 14                 1 0.421052631578947
    ## 15                 1 0.421052631578947
    ## 16                 1 0.473684210526316
    ## 17                 1 0.421052631578947
    ## 18                 1 0.421052631578947
    ## 19               0.5 0.473684210526316
    ## 20 0.666666666666667 0.473684210526316
    ## 21 0.333333333333333 0.473684210526316
    ## 22 0.333333333333333 0.473684210526316
    ## 23 0.666666666666667 0.473684210526316
    ## 24               0.5 0.421052631578947
    ## 25                 1 0.421052631578947
    ## 26                 1 0.421052631578947
    ## 27 0.666666666666667 0.473684210526316
    ## 28 0.666666666666667 0.473684210526316
    ## 29 0.666666666666667 0.473684210526316

``` r
ModelPerf[[2]]#Overall predictive performance. r2 for regression and MCC for classification
```

    ## [1] 0.7873563

## Plotting

``` r
VI <- mrVip(yhats, X=X) 
plot_vi(VI=VI,
        X=X,
        Y=Y,
        modelPerf=ModelPerf, 
       cutoff= 0, mode='classification') #The cutoff reduces the number of individual models printed in the second plot. 
```

![](README_files/figure-gfm/unnamed-chunk-7-1.png)<!-- -->

    ## Press [enter] to plot individual variable importance summaries

![](README_files/figure-gfm/unnamed-chunk-7-2.png)<!-- -->

## Effect of a feature on genetic change

We also wrap some flashlight functionality to visualize the marginal
(i.e. partial dependencies) or conditional (accumulated local effects)
effect of a feature on genetic change. Partial dependencies take longer
to calculate and are more sensitive to correlated features

``` r
flashlightObj <- mrFlashlight(yhats,
                              X=X,
                              Y=Y,
                              response = "single",
                              index=1,
                              mode='classification')

#plot prediction scatter for all responses. Gets busy with 
plot(light_scatter(flashlightObj,
                   v = "Forest",
                   type = "predicted"))
```

![](README_files/figure-gfm/unnamed-chunk-8-1.png)<!-- -->

``` r
#plots everything on one plot (partial dependency, ALE, scatter)
plot(light_effects(flashlightObj,
                   v = "Grassland"),
                   use = "all")
```

![](README_files/figure-gfm/unnamed-chunk-8-2.png)<!-- -->

``` r
#profileData_pd <- light_profile(flashlightObj,  v = "Grassland")

#mrProfileplot(profileData_pd , sdthresh =0.05) #sdthresh removes responses from the first plot that do not vary with the feature

profileData_ale <- light_profile(flashlightObj,
                                 v = "Grassland",
                                 type = "ale") #accumulated local effects

mrProfileplot(profileData_ale,
              sdthresh =0.01)
```

![](README_files/figure-gfm/unnamed-chunk-8-3.png)<!-- -->

    ##  Press [enter] to continue to the global summary plot

![](README_files/figure-gfm/unnamed-chunk-8-4.png)<!-- -->

``` r
#the second plot is the cumulative turnover function
```

## Interacting predictors or features

Finally, we can assess how features interact overall to shape genetic
change. Be warned this is memory intensive. Future updates to this
package will enable users to visualize these interactions and explore
them in more detail using 2D ALE plots for example.

``` r
#interactions <-mrInteractions(yhats,X,Y,mod='classification') #this is computationally intensive so multicores are needed. If stopped prematurely - have to reload things
#mrPlot_interactions(interactions,X,Y,top_ranking = 2,top_response=2)
```

## References

Xing, L, Lesperance, ML and Zhang, X (2020). Simultaneous prediction of
multiple outcomes using revised stacking algorithms. Bioinformatics, 36,
65-72. <doi:10.1093/bioinformatics/btz531>.

Fitzpatrick, M.C. & Keller, S.R. (2015) Ecological genomics meets
community-level modelling of biodiversity: mapping the genomic landscape
of current and future environmental adaptation. Ecology Letters 18,
1–16.doi.org/10.1111/ele.12376

Ellis, N., Smith, S.J. and Pitcher, C.R. (2012), Gradient forests:
calculating importance gradients on physical predictors. Ecology, 93:
156-168. <doi:10.1890/11-0252.1>


NEWS.md
mrIML 1.0.1
=======================================
* Initial release to github
* Initial functions and vignettes

mrIML 2.0.1
=======================================
* New functionality to allow MrIML to perform joint species distribution models (JSDMs)
* New bootstrap functions and vignette 
* Can also harness deep neural network models

R/pipeline.R
#'Testing functions out on real data.I will turn this into a vignette
#' to go with the paper. Currently we have two test datasets. 
#'One coinfection data set where columns represent parasite species.
#' Rows reflect sites. This dataset has only one feature (scale prop.
#' Currently just set up for binary classification. But multiclass/regression 
#' would be useful additions and easy
#'in the tidymodel environment. Multiclass could be pick SNP identities which would be useful.
#'
#
#lean toward mrIML as a name for the package.

#I'm thinking for the paper having three test datasets plus simulations.
#1. FIV bobcat data (host/pathogen example) 
#2. Amphibian SNP data (landscape genetics).
#3. Moose microbiome data. This would hit the key audience for Molecular Ecology resources.
#4.PRRS data on multi-strain circulation in swine https://www.biorxiv.org/content/10.1101/2020.04.09.034181v2

#There then could be separate papers on confection data extending 
#the approach to focus on co-occurence  (Lion/avian malaria data) 
#as well as a paper tracking between farm transisiion (Gustavo's data)

#--------------------------------------------------------------------------------------------------------------
# Use the ipak function to install and load multiple R packages. Thanks to Steven Worthington for this code.
# checks to see if packages are installed. Install them if they are not, then load them into the R session. Only work for CRAN 
#packages. Details are provided for the others required.

ipak <- function(pkg){
  new.pkg <- pkg[!(pkg %in% installed.packages()[, "Package"])]
  if (length(new.pkg)) 
    install.packages(new.pkg, dependencies = TRUE)
  sapply(pkg, require, character.only = TRUE)
}
#need a way to autoupdate or not each package - nice idea but not working

packages <- c('vip', 'mice', 'tidymodels', 'tidyverse', 'pdp', 'randomForest', 'caret', 'missForests', 'gbm', 'xgboost',
              'iml', 'doParallel',  'themis', 'viridis','janitor', 'hrbrthemes', 'ggrepel')

ipak(packages)
#--------------------------------------------------------------------------------------------------------------



#load packages. Reduce this list.
library(vip)
library(mice)
#library(VIM)
#library(imputeTS)
#library(fastshap)
library(tidymodels)
library(pdp)
library(randomForest)
library(caret)
#library(pROC)
#library(ROCR)
library(missForest)
library(gbm)
#library(iml)
library(tidyverse)
#library(parallel)
library(doParallel)
library(themis)
library(viridis)
library(janitor)
library(hrbrthemes)
library(MRFcov)
library(xgboost)
library(vegan)
library(ggrepel)
library(LEA) 
#if (!requireNamespace("BiocManager", quietly = TRUE))
# install.packages("BiocManager")

#BiocManager::install("LEA")
library(ape)
library(flashlight)
library(devtools)
install_github("mayer79/flashlight") 
#library(pbapply)

# load all function codes. This will disappear when we formally make this a function
source("./R/filterRareCommon.R")
source("./R/MrTidyModels.R")
source("./R/StackPredictions.R")
source("./R/devianceResids.R")
source("./R/filterRareCommon.R")
source("./R/MrtTidyPerf.R")
source("./R/mrvip.R")
source("./R/plot_vi.R")
#source("./R/mrPdP.R")

#new interaction code
source("./R/mrInteractions.R")
source("./R/mrPlotInteractions.R") #not finding this for some reason - no idea why
source("./R/vintTidy.R")

#Nick C - this and the function below are the functions with tidy model code I made from your original
source("./R/stacked_preds.R") #this does the stacking
source("./R/response_covariance") #this should create the covatiance matrix #not working?

source("./R/readSnpsPed.R") #function for reading SNP data from plink .ped file
source("./R/ResistanceComponents.R") #function for generating resistance component data from resistance matrices

source(("./R/mrFlashlight.R"))
source("./R/mrALEplots.R")
#----------------------------------------------------------------------------------------------------------------------

#---------------------------------------------------------------------------------
#Example: Bobcat SNP data- from Plink file
#---------------------------------------------------------------------------------

snps <- readSnpsPed("bobcat.plink.ped", "bobcat.plink.map.map") #NAs in data and interpolated as the mode. 
X <- filterRareCommon (snps, lower=0.4, higher=0.7) #these are harsh

X <- X[-c(279:281),] #these didnt match

#whatever individual X128 is it needs to go as it is a crazy outlier!

#filter correlated SNPS

df2 <- cor(X) #find correlations
hc <-  findCorrelation(df2, cutoff=0.5) # put any value as a "cutoff". This is quite high
hc <-  sort(hc)

 Labelled Y for simplicity

#Create resistance components for features using PCoA

#load resistance matrix generated using circuitscape. Add any matrix to a sub file in the working directory
#and provide that name to the function below. This will create a resistance component dataframe. Could add Mems here too

Y <- resist_components(data_resist, p_val=0.01,  filename = 'Bobcat_cs_matrices')

#check for correlations and remove
df2 <- cor(Y) #find correlations
hc <-  findCorrelation(df2, cutoff=0.9) # put any value as a "cutoff". This is quite high
hc <-  sort(hc)

Y <-  Y[,-c(hc)] #reduced set of Y features. Labelled Y for simplicity

#---------------------------------------------------------------------------------
#Example: Puma SNP data
#--------------------------------------------------------------------------------

snps <- readSnpsPed("ws.ped", "ws.map") #NAs in data and interpolated as the mode. 
X <- filterRareCommon (snps, lower=0.4, higher=0.75) #these are harsh

#---------------------------------------------------------------------------------
#Example: Bobcat FIV data  - read  viral data already curated
#---------------------------------------------------------------------------------

#load data for training
Responsedata <-read.csv("GF SNPs posBinary.csv")
rownames(Responsedata) <-Responsedata$X
Responsedata$X <- NULL
Responsedata[Responsedata==4] <-0 #fix random 4s in response data. 

#feature set
Features <-read.csv("Landscape and host data.csv", row.names = 1, head=T)
#summary(Features)

# instead we can devlop a cut % for interpolation in the func_load_data. Seems to be a random row somehow added to the feature data.
#Features<-na.interpolation(Features, option = "spline")

# # or remove NAs from the feature/predictor data.
FeaturesnoNA<-Features[complete.cases(Features), ];str(Features) #dropping NAs

Y <- FeaturesnoNA #for simplicity

#for more efficent testing for interactions (more variables more interacting pairs)
Y <- FeaturesnoNA[c(1:3)]
#Optional: Filter rare/common SNPs or species. Retaining minor allelle frequncies >0.1 and removing common allelles (occur>0.9)
fData <- filterRareCommon (Responsedata, lower=0.4, higher=0.7) 
X <- fData #for simplicity when comparing

#that occur in less than 35% of individuals and > 75% of individuals
#fData <- rownames_to_column(fData, "Individual")#get individual id back
#user can specify any tidy model here. 

#---------------------------------------------------------------------------------
#coinfection test data # from MRFcov. This data can also be used in this pipeline replacing X/Y
#---------------------------------------------------------------------------------

X <- select(Bird.parasites, -scale.prop.zos) #response variables eg. SNPs, pathogens, species....
Y <- select(Bird.parasites, scale.prop.zos) # feature set

#---------------------------------------------------------------------------------
#Simulated data
#---------------------------------------------------------------------------------
X <- read.csv('grid101.csv', row.names = NULL, head=T)
X[1:3] <- NULL
X[2:6] <- NULL; X[1:2] <- NULL #removing these for the moment.

#X[X==2] <- 1 #'cough' this is terrible  - it would be easier to have each loci as two columns to make this properly 
 #a classification model (rather than dealing with nominal data but i need to rethink probability)

#remove NAs
X[is.na(X)] <- 0 #not optimal. Will fix

#heavy filtering
#Xsim <- filterRareCommon (X, lower=0.2, higher=0.7)  #this isnt working for mutinomial

#check for correlations and remove. In this case all independent which makes sene
df2 <- cor(X, method = c('spearman')) #find correlations
df2[is.na(df2)] <- 0 #weird NAs in the last row
hc <-  findCorrelation(df2, cutoff=0.3) # put any value as a "cutoff".  removes half the SNPs. Aim for 50%
hc <-  sort(hc)

X <-  X[,-c(hc)] 

Y <- read.csv('simple_sims_env.csv')

#note that for multinomial models we do not have a great solution yet to calculate model deivance
#this means that model stacking currently is unavialable for this class of models.

#multinominal models arent working either...

#--------------------------------------------------------------------------------------------------------------------------------------
#creating the models

#-------------------------------------------------------------------
#Pre-training data visualization- checks for long tall distributions
#-------------------------------------------------------------------
## GM add here



#-------------------------------------------------------------------
#Set up the model
#-------------------------------------------------------------------

#user provides the models they would like for each component (model1)
 
model1 <- #model used to generate yhat
  # specify that the model is a random forest
  logistic_reg() %>%
  # select the engine/package that underlies the model
  set_engine("glm") %>%
  # choose either the continuous regression or binary classification mode
  set_mode("classification") #just for your response

model1 <- 
  rand_forest(trees = 100, mode = "classification") %>% #this should cope with multinomial data alreadf
  set_engine("ranger", importance = c("impurity","impurity_corrected")) %>%
  set_mode("classification")

#Boosted model (xgboost). Not tuned either currently. This doesn't work with the small bobcat dataset. Much too small

model1 <- 
boost_tree() %>%
  set_engine("xgboost") %>%
  set_mode("classification")

#for SVM need different tuning paramters. Currently tuning works best for tree-based algorithms.
  
#---------------------------------------------------------------------
#Perform the analysis
#---------------------------------------------------------------------  
#parallell processing
all_cores <- parallel::detectCores(logical = FALSE)

cl <- makePSOCKcluster(all_cores)
registerDoParallel(cl)

#models just using features/predictor variables.
yhats <- mrIMLpredicts(X=X,Y=Y, model1=model1, balance_data='no') ## in MrTidymodels. Balanced data= up updamples and down downsampled to create a balanced set

#-------------------------------------------------------------------
# Visualization for model tunning and performance
#-------------------------------------------------------------------
## GM add here

#we can now assess model performance from the best tuned model
ModelPerf <- mrIMLperformance(yhats, model1, X=X) # MCC is useful (higher numbers = better fit) 

ModelPerf[[1]] #predictive performance for individual responses 
ModelPerf[[2]]#overall predictive performance


#-------------------------------------------------------------------
# Visualization individual and global feature importance
#-------------------------------------------------------------------
## GM add here

#we can look at variable importance. Co-infection data only has one feature so not much use there.
VI <- mrVip(yhats, Y=Y) 
#plot model similarity

#plot variable importance

#for interpretation group features. Annoying but I cant see an easier way. Features are in alphabetical order
groupCov <- c(rep ("Host_characteristics", 1),rep("Urbanisation", 3), rep("Vegetation", 2), rep("Urbanisation",1), rep("Spatial", 2), 
              rep('Host_relatedness', 6),rep ("Host_characteristics", 1),rep("Vegetation", 2), rep("Urbanisation",1))  

#not that GLMs in particular wont produce coefficents for features that are strongly colinear and will drop them from the model.
#in this case group cov will have to be changed to reflect features included in the model. 

plot_vi(VI=VI,  X=X,Y=Y, modelPerf=ModelPerf, groupCov=groupCov, cutoff= 0.5, plot.pca='no')#note there are two plots here. PCA is hard to read with > 50 response varianbles

#if you dont want/need to group covariates:

plot_vi(VI=VI,  X=X,Y=Y, modelPerf=ModelPerf, cutoff= 0, plot.pca='yes') #mcc cutoff not working right

#First plot is overall importance, the second a pca showing responses with similar importance scores and the third is individual SNP models.
#warning are about x axis labels so can ignore 


# can build a flaslight object for individual responses 

fl <- mrFlashlight(yhats, X, Y, response = "single", index=1)

plot(light_performance(fl), fill = "orange") +
  
  labs(x = element_blank())

str(fl)

#plot(light_breakdown(fl , new_obs = cbind(X, Y)[1, ]),by = X, v=Y) #prints all responses - need to fix but could be quite handy.

#int <- light_interaction(fl, pairwise=TRUE) #not working, but possible!

#------------------------------------------------------------
#Multiple response

flashlightObj <- mrFlashlight(yhats, X, Y, response = "multi")

#plots


plot(light_profile(flashlightObj, v = "broad60ibd_resistances_Axis.1", type = "ale"))
#this will plot all responses. See mrALEplots below

#another way to assess performance for each response. Lots of response make it hard to read
plot(light_performance(multifl), rotate_x = TRUE, fill = "orange") +
  
  labs(x = element_blank())


#plot prediction scatter for all responses. Gets busy with 
plot(light_scatter(flashlightObj, v = "Grassland", type = "predicted"))

#plots everything on one plot (partial dependency, ALE, scatter)
plot(light_effects(flashlightObj, v = "Grassland"), use = "all")


profileData_pd <- light_profile(flashlightObj, v = "Grassland") #partial dependencies
profileData_ale <- light_profile(flashlightObj, v = "Grassland", type = "ale") #acumulated local effects

#Plot global ALE plot. This the plots the smoothed average ALE value. Have to remove responses that don't respond.

mrProfileplot(profileData_pd , sdthresh =0.05)
mrProfileplot(profileData_ale , sdthresh =0.05)


#-------------------------------------------------------------------------------------------------

#calculate interactions  -this is qute slow and memory intensive

interactions <-mrInteractions(yhats, X, Y) #this is computationally intensive so multicores are needed. If stopped prematurely - have to reload things

mrPlot_interactions(Interact, X,Y, top_ranking = 3, top_response=3)


#-------------------------------------------------------------------------------------------------


#adding other response varables to see if this improves predictions. wecould say that it only has been tested 
#on 3 algorithms (GAMs, Xgboost which is an improved GBM and liner models) and user beware otherwise.

covariance_mod <- 
  boost_tree() %>%
  set_engine("xgboost") %>%
  set_mode("regression")

#or 
covariance_mod <- 
  rand_forest(trees = 100, mode = "regression") %>%
  set_engine("ranger", importance = "permutation") %>%
  set_mode("regression")

#rhats <- StackPredictions(yhats, data.test, covariance_mod = 'gam')

#calculate covariance predictions. Nick - I just split this function up as I imagine it would be handy.
#at this bit,
rhats_2 <- response_covariance(yhats, covariance_mod)

#stack everything together
finalPred <- stacked_preds(rhats_2, yhats)

#add similarity in predictions to look for SNP patterns.

#at some point it would be good to add interactions/Shapely here too. 

#---------------------------------------------------------------------------------------------------------------

#Extra importance plots:

#another way to look at it
ModelPerf_p<-do.call(rbind.data.frame, ModelPerf)

## perfromance by outcome
ModelPerf_p%>%
  drop_na()%>%
  ggplot(aes(sensitivity, specificity, shape=response, colour=response  , fill=response)) +
  geom_smooth(method="lm") +
  geom_point(size=3)

## also
ModelPerf_p%>%
  drop_na()%>%
  ggplot(aes(mcc, specificity, shape=response, colour=response, fill=response   )) +
  geom_smooth(method="lm") +
  geom_point(size=3)

### Not sure we want to have this
ModelPerf_p[3] #summary mcc for all response variables
